add support for sen66

ackPeng opened the pull request:
# What does this implement/fix?

Added support for sen66 sensors based on sen5x

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- fixes <[link to issue](https://github.com/esphome/feature-requests/issues/2995)>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#<https://github.com/esphome/esphome-docs/pull/4710>

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
i2c:
  - id: i2c_sen5x
    scl: GPIO7
    sda: GPIO6

sensor:
  - platform: sen5x
    id: sen54
    temperature:
      name: Temperature
      accuracy_decimals: 1
    humidity:
      name: Humidity
      accuracy_decimals: 0
    pm_1_0:
      name: PM <1µm Weight concentration
      id: pm_1_0
      accuracy_decimals: 1
    pm_2_5:
      name: PM <2.5µm Weight concentration
      id: pm_2_5
      accuracy_decimals: 1
    pm_4_0:
      name: PM <4µm Weight concentration
      id: pm_4_0
      accuracy_decimals: 1
    pm_10_0:
      name: PM <10µm Weight concentration
      id: pm_10_0
      accuracy_decimals: 1
    co2:
      name: CO2
      accuracy_decimals: 1
    nox:
      name: NOx
    voc:
      name: VOC
      algorithm_tuning:
        index_offset: 100
        learning_time_offset_hours: 12
        learning_time_gain_hours: 12
        gating_max_duration_minutes: 180
        std_initial: 50
        gain_factor: 230
    temperature_compensation:
      offset: 0
      normalized_offset_slope: 0
      time_constant: 0
    auto_cleaning_interval: 604800s
    acceleration_mode: low
    store_baseline: true
    address: 0x6b
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

ackPeng committed:
add support for sen66

ackPeng committed:
add sen66 test

ackPeng committed:
change some formatting

ackPeng committed:
Merge branch 'esphome:dev' into dev

ackPeng committed:
Merge branch 'esphome:dev' into dev

npfusaro reviewed (changes requested):
This doesn't add support as much as it replaces the support for the current SEN5x series and allows the SEN66 to work with the same files. By adding this code everyone who uses the current SEN55 devices code will no longer work. As the SEN6X series is a completely new platform with new sensors incorporated this should be its own component.

ackPeng commented:
@npfusaro  The newly added support for SEN66 does not affect the functions of the SEN5X series. I just expanded the SEN66 related instructions based on SEN5X (in fact, most of the instructions of the two are the same)

npfusaro commented:
@ackPeng apologies. I now see that you refactored some of the names to be less specific. On first read through I only noticed the that you removed 5x commands and added 6x .I agree that the command sets are mostly identical. Its just that I think the naming of the platform could cause a bit of confusion.

ackPeng committed:
Merge branch 'esphome:dev' into dev

ackPeng commented:
@npfusaro The naming of the library can indeed cause confusion; however, since the instruction sets of the SEN66 and SEN5x are largely the same, their code is also very similar. If we were to create a separate library specifically for SEN6x, the resulting code would be extremely similar. This seems to go against ESPHome's code review guidelines (as far as I know, ESPHome does not allow highly similar code to be PR'd into the main branch). Perhaps the best solution is to change the library's name? Do you have any suggestions in this regard?

ackPeng committed:
Merge branch 'esphome:dev' into dev

realeol commented:
Something I already missed with the SEN5x-Integration was the option to read the raw sensor values. According to the Sensation Datasheets of the SEN5x as well as the SEN6x, it is possible (see pages 36 / 37 in datasheet for the SEN6x / dedicated document for the SEN5x). While some values are not useful, the number concentration for PM as well as the raw Output of the MOX-Sensor could be useful to some people, I'd imagine. Unfortunately, I myself am not really good at coding, so I can't provide much help.

krystof-k commented:
Just a quick observation: I just tested this with SEN66 and it seems to work just fine.

cracrama commented:
SEN 66 has atmospheric pressure sensor according Datasheet. If that could be exposed that would be great addition.

daradib commented:
I tested this on SEN66. Looks good aside from two issues:

1. Fan cleaning does not work, neither automatically nor manually. `auto_cleaning_interval` (0x8004) is not supported on SEN6x. And according to the data sheet, `start_fan_autoclean` (0x5607) is only available in idle mode, unlike SEN5x. So measurement would need to be stopped to fix this.

2. VOC and NOx are initially reported for a few seconds as 3276.7. Temperature, Humidity, VOC, and NOx are signed integers and 0x7FFF values should be parsed as NAN, similar to how `sen55.cpp` currently handles 0xFFFF values.

Also, `acceleration_mode` is not supported on SEN6x, though there is a different command with filter constants (0x6100) that could added in the future if need be.

daradib commented:
#8769 should fix the issue with signed integer null values which also affects SEN5x.

npfusaro commented:
> SEN 66 has atmospheric pressure sensor according Datasheet. If that could be exposed that would be great addition.

I have made a PR that exposes the Barometric pressure sensor data. Keep in mind that this sensor is only useful as every time the unit resets it defaults to 1013 hPa. There is a way to set the current pressure but it does not store the anything after a reset. It is an odd choice by sensorion to include the pressure sensor sensor to help with the CO2 but always reset to 1013 on start. It would be one thing to have a value that drifts or need calibration but if the ambient was 9950 or 1013 it still defaults to 1013.

daradib commented:
@cracrama @npfusaro Are you sure the pressure sensor is included in the SEN66? I think the pressure/altitude commands require an external sensor (not included) for pressure compensation, as is done by the [SCD4x component](https://esphome.io/components/sensor/scd4x.html).

npfusaro commented:
> @cracrama @npfusaro Are you sure the pressure sensor is included in the SEN66? I think the pressure/altitude commands require an external sensor (not included) for pressure compensation, as is done by the [SCD4x component](https://esphome.io/components/sensor/scd4x.html).

You are correct. I took a closer look at the data sheet and see that this is infact the same as the scd40. I thought I had read it was included in the temp/humidity ic similar to the bme280. The pressure on this device is used for compensation on the CO2 sensor.

disruptivepatternmaterial commented:
I used the config in this thread and the pr in my local esphome with a sen66 and it is mostly working fine except for the noted issues in this pr, it is reading the same numbers as other sensors in the same box. So well done!

disruptivepatternmaterial commented:
I take it back :) and hope this helps...know how people like to be so kind on the internet, I am just trying to contribute a little where I can!

```
  - platform: sen5x
    id: sen55
    pm_1_0:
      name: "PM <1µm Weight concentration"
      id: PM_1_0
      accuracy_decimals: 2
      force_update: true
    pm_2_5:
      name: "PM <2.5µm Weight concentration"
      id: PM_2_5
      accuracy_decimals: 2
      force_update: true
    pm_4_0:
      name: "PM <4µm Weight concentration"
      id: PM_4_0
      accuracy_decimals: 2
      force_update: true
    pm_10_0:
      name: "PM <10µm Weight concentration"
      id: PM_10_0
      accuracy_decimals: 2
      force_update: true
    temperature:
      name: "SEN55 Temperature"
      id: sen55_temp
      accuracy_decimals: 2
      internal: true
      force_update: true
    humidity:
      name: "SEN55 Humidity"
      id: sen55_humid
      accuracy_decimals: 2
      internal: true
      force_update: true
    voc:
      name: VOC
      id: voc
      accuracy_decimals: 2
      force_update: true
      algorithm_tuning:
        index_offset: 100
        learning_time_offset_hours: 12
        learning_time_gain_hours: 12
        gating_max_duration_minutes: 180
        std_initial: 50
        gain_factor: 230
    co2:
      name: CO2
      accuracy_decimals: 1
    nox:
      name: NOX
      id: nox
      accuracy_decimals: 2   
      force_update: true   
      algorithm_tuning:
        index_offset: 100
        learning_time_offset_hours: 12
        learning_time_gain_hours: 12
        gating_max_duration_minutes: 180
        std_initial: 50
        gain_factor: 230
    temperature_compensation:
      offset: 0
      normalized_offset_slope: 0
      time_constant: 0
    acceleration_mode: low
    store_baseline: true
    update_interval: $sensor_interval
    address: 0x6b

```

I am using the above and it appears that all the details under NOX/algorithm_tuning, which work find on my sen5x version (I have multiple AiqQ's from M5 which report NOX). What is not happening, in particular is the mapping to 100 vs. 1 so that I cannot use the same normal code to report the word version of the status (improving, normal, etc...).

Again hope this helps, if I have time I will see if I can figure out what might be happening - as I am not 100% that the 6 series work the same way ask the 5 series as related to NOX and VOC.

npfusaro commented:
> I take it back :) and hope this helps...know how people like to be so kind on the internet, I am just trying to contribute a little where I can!
> 
> ```
>   - platform: sen5x
>     id: sen55
>     pm_1_0:
>       name: "PM <1µm Weight concentration"
>       id: PM_1_0
>       accuracy_decimals: 2
>       force_update: true
>     pm_2_5:
>       name: "PM <2.5µm Weight concentration"
>       id: PM_2_5
>       accuracy_decimals: 2
>       force_update: true
>     pm_4_0:
>       name: "PM <4µm Weight concentration"
>       id: PM_4_0
>       accuracy_decimals: 2
>       force_update: true
>     pm_10_0:
>       name: "PM <10µm Weight concentration"
>       id: PM_10_0
>       accuracy_decimals: 2
>       force_update: true
>     temperature:
>       name: "SEN55 Temperature"
>       id: sen55_temp
>       accuracy_decimals: 2
>       internal: true
>       force_update: true
>     humidity:
>       name: "SEN55 Humidity"
>       id: sen55_humid
>       accuracy_decimals: 2
>       internal: true
>       force_update: true
>     voc:
>       name: VOC
>       id: voc
>       accuracy_decimals: 2
>       force_update: true
>       algorithm_tuning:
>         index_offset: 100
>         learning_time_offset_hours: 12
>         learning_time_gain_hours: 12
>         gating_max_duration_minutes: 180
>         std_initial: 50
>         gain_factor: 230
>     co2:
>       name: CO2
>       accuracy_decimals: 1
>     nox:
>       name: NOX
>       id: nox
>       accuracy_decimals: 2   
>       force_update: true   
>       algorithm_tuning:
>         index_offset: 100
>         learning_time_offset_hours: 12
>         learning_time_gain_hours: 12
>         gating_max_duration_minutes: 180
>         std_initial: 50
>         gain_factor: 230
>     temperature_compensation:
>       offset: 0
>       normalized_offset_slope: 0
>       time_constant: 0
>     acceleration_mode: low
>     store_baseline: true
>     update_interval: $sensor_interval
>     address: 0x6b
> ```
> 
> I am using the above and it appears that all the details under NOX/algorithm_tuning, which work find on my sen5x version (I have multiple AiqQ's from M5 which report NOX). What is not happening, in particular is the mapping to 100 vs. 1 so that I cannot use the same normal code to report the word version of the status (improving, normal, etc...).
> 
> Again hope this helps, if I have time I will see if I can figure out what might be happening - as I am not 100% that the 6 series work the same way ask the 5 series as related to NOX and VOC.



Remove the line index_offset: 100 under algorithm tuning. That will return it to standard sen55 starting index of 1.

jesserockz commented on esphome/components/sen5x/sen5x.cpp:
```diff
@@ -97,36 +117,35 @@ void SEN5XComponent::setup() {
+      } else if (product_name_ == "SEN55") {
+        sen5x_type = SEN55;
+      } else if (product_name_ == "SEN66" ||
+                 this->co2_sensor_) {  // Reading the product name of sen66 will return empty, so the sensor type is
```
I dont think this is necessary, people might set it accidentally not knowing it is not supported by the chip they have.

mikelawrence replied:
I think this was a sneaky way of adding the CO2 sensor if the product name responded with an empty string. My SEN66 reports correctly.

jesserockz commented on esphome/components/sen5x/sen5x.cpp:
```diff
@@ -97,36 +117,35 @@ void SEN5XComponent::setup() {
-      }
-      if (this->nox_sensor_ && sen5x_type != SEN55) {
+      } else if (this->nox_sensor_ && sen5x_type != SEN55 && sen5x_type != SEN66) {
         ESP_LOGE(TAG, "For NOx a SEN55 is required. You are using a <%s> sensor", this->product_name_.c_str());
```
Log will need to change to say sen66 too

jesserockz commented on esphome/components/sen5x/sen5x.cpp:
```diff
@@ -324,15 +344,24 @@ void SEN5XComponent::update() {
+    uint16_t measurements[9];
 
-    if (!this->read_data(measurements, 8)) {
+    if (!this->read_data(measurements, 9)) {
```
Will this read correctly on non sen66?

mikelawrence replied:
No it will not. The SEN5X sensors each return a compatible set of values based on capabilities. The SEN6x models required parsing based on the model.

mikelawrence commented:
This PR is missing support for SEN6X sensors other than SEN66.

I have modified this PR to support all SEN6X models while keeping the SEN5X models intact but I am not sure how to make the changes in this PR. 

Notes:
* SEN6X models do not support Auto Fan Cleaning.
* SEN6X models support SHT4X heater in a limited way (at least from a SHT4X device).
* SEN6X Temperature compensation is different from SEN5x. Sensirion has not released a promised Application Note on how to use Temperature Compensation on the SEN6X models yet. So Temperature Compensation for these models included. Datasheet is not enough to figure this out.

Additions:
* More model logic. Note the Sen60 is only PM and changes every I2C command due the the lack of an embedded micro.
* HCHO Formaldehyde sensor for the SEN68 model.
* CO2 auto_self_calibration, altitude_compensation and ambient_pressure_compensation_source configuration.
* set_ambient_pressure_compensation and perform_forced_co2_calibration Actions for CO2 sensor
* activate_heater Action for Humidity sensor

These changes are on [Github here](https://github.com/mikelawrence/ESPHome-Components/tree/main/components/sen5x).

Readme [here](https://github.com/mikelawrence/ESPHome-Components?tab=readme-ov-file#sen5x-external-component).

jesserockz commented:
@mikelawrence There are 3 methods to get your code into ESPHome:
1. Make a PR into @ackPeng branch on their repo, once merged, the changes will show up in this PR
2. Wait until this PR is merged then make a further one
3. Make an entire new PR incorporating all changes (not as nice as it removes recognition of original author of this PR)

@ackPeng Please review @mikelawrence's changes and let me know what you would prefer to do from here

Jesse

Please ping me in discord in #devs if I do not respond in a timely fashion here

mikelawrence committed:
Support all SEN6X models

Added missing support for SEN6X sensors other than SEN66. Only the SEN66 is available at this time, but the SEN60, SEN63C, SEN65 and SEN68 are scheduled to be released this year.

Notes:
* SEN6X models do not support Auto Fan Cleaning.
* SEN6X models support SHT4X heater in a limited way (at least from a SHT4X device).
* SEN6X Temperature compensation is different from SEN5x. Sensirion has not released a promised Application Note on how to use Temperature Compensation on the SEN6X models yet. So Temperature Compensation for these models is NOT included. Datasheet is not enough to figure this out.

Additions:
* More model logic. Note the Sen60 is only PM and changes every I2C command due the the lack of an embedded micro.
* model configuration enumeration supporting SEN50, SEN54, SEN55, SEN60, SEN63C, SEN65, SEN66 and SEN68. I would think it's better to REQUIRE a model in the configuration so the configuration could be validated at install time instead of runtime, but as optional it won't break current sen5x configurations. It also supports a blank SEN66 product name string mentioned in the original PR. Note I have tested two SEN66 sensors and neither responded with a blank product name string.
* HCHO Formaldehyde sensor. New for the SEN68 model.
* CO2 auto_self_calibration, altitude_compensation and ambient_pressure_compensation_source configuration.
* set_ambient_pressure_compensation and perform_forced_co2_calibration Actions for CO2 sensor. New for SEN63C and SEN66 sensors.
* activate_heater Action for Humidity sensor. New for SEN63C, SEN65, SEN66 and SEN68 sensors.

mikelawrence committed:
Correct action return values and README cleanup

ackPeng committed:
Merge pull request #2 from mikelawrence/dev

Support all SEN6X models

ackPeng commented:
@mikelawrence Sorry for the late reply. I was busy recently and didn't pay attention to it in time. You can submit a new PR

disruptivepatternmaterial commented:
booooooooooo

mikelawrence commented:
@disruptivepatternmaterial There is another PR based on this one waiting for approval.

PR #9254
